File: apps/ios/Barns/BarnsApp/Presentation/MyItems/EditGreeneryViewModel.swift
import Combine
import Foundation

/// Form state for editing an existing registered greenery (`ProductItem`).
/// Local-only: edits are kept in this presentation state until the user
/// explicitly saves; cancelling simply discards this view model.
@MainActor
final class EditGreeneryViewModel: ObservableObject {
    @Published var name: String
    @Published var type: ProductItemType
    @Published var locationLabel: String
    @Published var notes: String
    @Published var status: ProductItemStatus
    @Published private(set) var isSaving = false
    @Published private(set) var errorMessage: String?

    private let original: ProductItem
    private let updateProductItemUseCase: UpdateProductItemUseCase

    init(item: ProductItem, updateProductItemUseCase: UpdateProductItemUseCase) {
        self.original = item
        self.updateProductItemUseCase = updateProductItemUseCase
        name = item.name
        type = item.type
        locationLabel = item.locationLabel ?? ""
        notes = item.notes ?? ""
        status = item.status
    }

    var canSave: Bool {
        !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && !isSaving
    }

    /// Updates the item locally. Returns true on success so the view can dismiss.
    func save() async -> Bool {
        guard canSave else { return false }
        isSaving = true
        errorMessage = nil
        defer { isSaving = false }
        do {
            _ = try await updateProductItemUseCase.execute(
                original: original,
                name: name.trimmingCharacters(in: .whitespacesAndNewlines),
                type: type,
                locationLabel: emptyToNil(locationLabel),
                notes: emptyToNil(notes),
                status: status
            )
            return true
        } catch {
            errorMessage = "Unable to save changes. Please try again."
            return false
        }
    }

    private func emptyToNil(_ value: String) -> String? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

File: apps/android/app/src/main/java/com/barns/app/presentation/myitems/EditGreeneryViewModel.kt
package com.barns.app.presentation.myitems

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.barns.app.domain.model.ProductItem
import com.barns.app.domain.model.ProductItemStatus
import com.barns.app.domain.model.ProductItemType
import com.barns.app.domain.usecase.myitems.UpdateProductItemUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Form state for editing an existing registered greenery (ProductItem).
 * Local-only: edits are kept in this presentation state until the user
 * explicitly saves; cancelling simply discards this view model.
 */
class EditGreeneryViewModel(
    private val item: ProductItem,
    private val updateProductItemUseCase: UpdateProductItemUseCase,
) : ViewModel() {
    private val _state = MutableStateFlow(
        EditGreeneryUiState(
            name = item.name,
            type = item.type,
            locationLabel = item.locationLabel ?: "",
            notes = item.notes ?: "",
            status = item.status,
        ),
    )
    val state: StateFlow<EditGreeneryUiState> = _state.asStateFlow()

    fun onNameChange(value: String) { _state.value = _state.value.copy(name = value) }
    fun onLocationChange(value: String) { _state.value = _state.value.copy(locationLabel = value) }
    fun onNotesChange(value: String) { _state.value = _state.value.copy(notes = value) }
    fun onTypeChange(value: ProductItemType) { _state.value = _state.value.copy(type = value) }
    fun onStatusChange(value: ProductItemStatus) { _state.value = _state.value.copy(status = value) }

    /** Updates the item locally and invokes [onSaved] on success. */
    fun save(onSaved: () -> Unit) {
        val current = _state.value
        if (!current.canSave) return
        _state.value = current.copy(isSaving = true, errorMessage = null)
        viewModelScope.launch {
            runCatching {
                updateProductItemUseCase.execute(
                    original = item,
                    name = current.name.trim(),
                    type = current.type,
                    locationLabel = current.locationLabel.trim().ifBlank { null },
                    notes = current.notes.trim().ifBlank { null },
                    status = current.status,
                )
            }
                .onSuccess {
                    _state.value = _state.value.copy(isSaving = false)
                    onSaved()
                }
                .onFailure {
                    _state.value = _state.value.copy(
                        isSaving = false,
                        errorMessage = "Unable to save changes. Please try again.",
                    )
                }
        }
    }
}

data class EditGreeneryUiState(
    val name: String = "",
    val type: ProductItemType = ProductItemType.INSTALLED,
    val locationLabel: String = "",
    val notes: String = "",
    val status: ProductItemStatus = ProductItemStatus.ACTIVE,
    val isSaving: Boolean = false,
    val errorMessage: String? = null,
) {
    val canSave: Boolean
        get() = name.isNotBlank() && !isSaving
}
